### ISSUE TITLE
Syntax highlighting breaks if the file is too big

### ISSUE THREAD

**Issue by perfecto25** [Microsoft/vscode#34240]
- VSCode Version: 1.11.2
- OS Version: centos 7

Steps to Reproduce:

1. open small (less than 10MB) file, type any text, add a comment (#blah blah) select syntax, should see color change according to syntax
2. open a large file (over 10MB), same steps, syntax doesnt highlight

Reproduces without extensions: Yes/No

Heres an example of a small file (Snort rules file) (1.1KB)

syntax highlighting is visible, comments are grayed out
![image](https://user-images.githubusercontent.com/10224886/30339535-5a2961e6-97bd-11e7-8098-f58b658a8984.png)

heres a large file (29MB)
setting syntax to Shell, does not gray out comments, 

![image](https://user-images.githubusercontent.com/10224886/30339693-dcb667d0-97bd-11e7-9d9c-5c1df6f4131b.png)

**Comment by alexandrudima** [comment]
This is currently as designed. We want to reduce our memory usage when opening large files. Representing tokens (currently each token takes 2*32bits in an array) is a big source of memory usage.

Currently, our strategy is the following:
* if a file is larger than 20MB or if it has more than 300K lines
* we reduce our feature set in order to save memory:
  * no colorization
  * no word wrapping 
  * no indent guides

Basically, we do (hopefully graceful) degradation of our feature set in order to reduce memory consumption... and not crash with an OOM exception :).

**Comment by perfecto25** [comment]
is there anyway to disable this? I can open this large file in other editors with highlight showing with no issues. Cant do it with VSC.  Editing large Snort rule files is very hard without syntax highlighting , hard to tell whats commented out or not.

**Comment by Jatz** [comment]
+1 for this suggestion. 

Would it be possible to add a setting (obviously disabled by default) to override this or at least increase the threshold somehow?

I'm currently trying to port over a syntax highlighter from Sublime Text into VSCode for parsing log files and traces produced by an ERP system we use at our workplace. I've pretty much got the plugin working well for files that are less than 20mb in size. It would be nice to be able to let the user decide whether they would like to enable syntax highlighting for large files, even if there was a warning that showed the user that the file was large and could take a while to parse. The plugin I wrote for Sublime Text works 
 well, albeit with a bit of lag if the file is excessively large, and that's a trade off most users of the plugin are willing to make since it makes reading a log/trace file so much easier.

**Comment by ecampidoglio** [comment]
While I understand the reasoning behind this choice, I think users should be warned about what's happening so that they don't open, say, a large JSON file and immediately start thinking that syntax highlighting is broken or—even worse—that the file has an invalid format. 🙁

VSCode does show a warning when trying to compare a large file:

![vscodelargefilewarning](https://user-images.githubusercontent.com/877348/34609570-a2d1662c-f21d-11e7-8122-dfb807872a17.png)

A similar warning saying something along the lines of:

> Some editor features have been disabled because the file is too large. Click here for more details.

would be very helpful.

**Comment by gusbemacbe** [comment]
VSCode is very similar to Atom due to be written in Electron. Different of Sublime Text which doesn't break the syntax highlighting and loads quickly the big file.

**Comment by chrisvio** [comment]
What a disappointing limitation. I have a 244**K** JSON file I can't work with due to this limitation. The extension I use to format JSON files won't work because of this. Tried other extensions. Same deal. I guess it's back to Sublime. The same thing happened when I tried Atom. A shame. I downloaded and installed VSCode a few days ago and was really digging it. Why not make this limitation something that is user configurable?

**Comment by gusbemacbe** [comment]
@chrisvio it does not happen only with VSCode, it happens with other apps written in Electron. Atom and VSCode are good only for small files and better for programming and snippets, while Sublime Text is good for big files and finding and replacing something all files once in a while and bad for programming and snippets.

**Comment by gusbemacbe** [comment]
@Jatz Be careful, Sublime Text is not open source.

**Comment by Jatz** [comment]
@gusbemacbe Sorry, I haven't been following this thread for a while. What exactly do I need to be careful with?

**Comment by gusbemacbe** [comment]
@Jatz, Sublime Text is a proprietary software and is licensed under EULA. It is not open source. It is not good idea to port EULA-licensed Sublime Text plugin to an open source app.

**Comment by Jatz** [comment]
@gusbemacbe Wow ok good to know. I'm hoping I just used the wrong semantics here. Maybe the word "port" is not the correct word. I wrote the Sublime Text plugin, and I've made similar functionality for VSCode but have completely rewritten the code (from scratch) for the VSCode plugin, so it shares absolutely no code in common at all.

**Comment by StarNamer** [comment]
I don't supposed there's any chance this could be made user selectable? I've currently got a 1.1 million line C# file open  (53Mb on disk) and the memory usage shown in Task Manager details in only 212Mb. The PC has 64Gb and I've happily run multiple instances of Visual Studio using more than 5Gb of memory.

From the 2*32bits/token value and assuming an average of 6 tokens per line (guess) turning it on for my 1.1 million line file would use about 53Mb, which I wouldn't think was a major problem.

**Comment by andre161292** [comment]
For me it seems to be already possible to deactivate this feature by setting "editor.largeFileOptimizations" to false. Correct me if i'm wrong. 
It still needs a longer time than Sublime to highlight the whole file, but for really large files (~ >500M) it loads faster.

**Comment by codeaid** [comment]
@andre161292 That worked! Thanks for the tip. It takes a bit for the whole 350,000 line `highcharts.d.ts` file to get coloured in but it's only a few seconds.
